File: feature/keyboard/src/main/java/net/twisterrob/colorfilters/android/keyboard/BaseKeyboardHandler.java
package net.twisterrob.colorfilters.android.keyboard;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.inputmethodservice.KeyboardView;
import android.os.Build;
import android.support.annotation.NonNull;
import android.text.*;
import android.util.Log;
import android.view.*;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

// http://www.fampennings.nl/maarten/android/09keyboard/index.htm
// http://forum.xda-developers.com/showthread.php?t=2497237
public abstract class BaseKeyboardHandler implements KeyboardHandler {

	private boolean haptic;
	protected final Context context;
	protected final Window window;
	protected final KeyboardView keyboardView;
	private CustomKeyboardListener listener;

	public BaseKeyboardHandler(Window window, KeyboardView keyboardView) {
		this.window = window;
		this.context = keyboardView.getContext().getApplicationContext();
		this.keyboardView = keyboardView;

		hideCustomKeyboard();
		window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
		keyboardView.setPreviewEnabled(false);
		setHapticFeedback(true);
	}

	public void hideCustomKeyboard() {
		keyboardView.setVisibility(View.GONE);
		keyboardView.setEnabled(false);
		clearFocus();
		if (listener != null) {
			listener.customKeyboardHidden();
		}
	}

	protected void clearFocus() {
		View focused = window.getCurrentFocus();
		if (focused != null) {
			focused.clearFocus();
		}
	}

	public void showCustomKeyboard(@NonNull View v) {
		InputMethodManager imm = (InputMethodManager)context.getSystemService(Activity.INPUT_METHOD_SERVICE);
		imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
		keyboardView.setEnabled(true);
		keyboardView.setVisibility(View.VISIBLE);
		if (listener != null) {
			listener.customKeyboardShown();
		}
	}

	public void registerEditText(@NonNull EditText editText) {
		// alternative to this is editText.setOnTouchListener(new OnTouchWrapper());
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
			editText.setInputType(InputType.TYPE_NULL);
		} else {
			try {
				editText.getClass().getMethod("setShowSoftInputOnFocus", boolean.class)
				        .invoke(editText, false); //NON-NLS
			} catch (Exception ex) {
				Log.w("HACK", "Could turn of input focus for EditText: " + editText, ex); //NON-NLS
			}
		}
		editText.setFocusable(true);
		editText.setFocusableInTouchMode(true);
		editText.setOnFocusChangeListener(new MyOnFocusChangeListener());
		editText.setOnClickListener(new MyOnClickListener());
	}

	public void unregisterEditText(@NonNull EditText editText) {
		editText.setOnFocusChangeListener(null);
		editText.setOnClickListener(null);
		editText.clearFocus();
	}

	public void setHapticFeedback(boolean haptic) {
		keyboardView.setHapticFeedbackEnabled(haptic);
		this.haptic = haptic;
	}

	public boolean handleBack() {
		if (keyboardView.handleBack()) {
			return true;
		}
		if (keyboardView.getVisibility() == View.VISIBLE) {
			hideCustomKeyboard();
			return true;
		}
		return false;
	}

	public void setCustomKeyboardListener(@NonNull CustomKeyboardListener listener) {
		this.listener = listener;
	}

	@SuppressWarnings("unused")
	@SuppressLint("ClickableViewAccessibility")
	private static class OnTouchWrapper implements View.OnTouchListener {
		@Override
		public boolean onTouch(View v, MotionEvent event) {
			EditText editText = (EditText)v;
			int backup = editText.getInputType();
			editText.setInputType(InputType.TYPE_NULL); // Disable standard keyboard
			try {
				return editText.onTouchEvent(event); // Call native handler
			} finally {
				editText.setInputType(backup);
				editText.setCursorVisible(true);
			}
		}
	}

	private class MyOnClickListener implements View.OnClickListener {
		@Override
		public void onClick(View v) {
			showCustomKeyboard(v);
		}
	}

	private class MyOnFocusChangeListener implements View.OnFocusChangeListener {
		@Override
		public void onFocusChange(View v, boolean hasFocus) {
			if (hasFocus) {
				showCustomKeyboard(v);
			} else {
				hideCustomKeyboard();
			}
		}
	}

	protected abstract class BaseOnKeyboardActionListener implements KeyboardView.OnKeyboardActionListener {
		protected static final int KEY_MOVE_LEFT = '\u2190';
		protected static final int KEY_MOVE_RIGHT = '\u2192';
		protected static final int KEY_MOVE_START = '\u21E4';
		protected static final int KEY_MOVE_END = '\u21E5';

		protected static final int KEY_DONE = '\u0017';
		protected static final int KEY_CLEAR = '\u0018';
		protected static final int KEY_BACKSPACE = '\u0008';

		protected View findView() {
			return window.getCurrentFocus();
		}

		protected EditText findEdit() {
			View focusCurrent = findView();
			if (focusCurrent instanceof EditText) {
				return (EditText)focusCurrent;
			}
			return null;
		}

		@Override
		public void onKey(int primaryCode, int[] keyCodes) {
			EditText editor = findEdit();
			if (editor == null) {
				hideCustomKeyboard();
				return;
			}

			Editable editable = editor.getText();

			// if there's a selection, clear it to make any edits replace actions
			int start = editor.getSelectionStart();
			int end = editor.getSelectionEnd();
			if (start < end) {
				editable.delete(start, end);
			}

			switch (primaryCode) {
				case KEY_CLEAR:
					if (editable != null) {
						editable.clear();
					}
					break;
				case KEY_BACKSPACE: {
					if (editable != null && 0 < start) {
						editable.delete(start - 1, start);
					}
					break;
				}
				case KEY_DONE:
					hideCustomKeyboard();
					break;
				case KEY_MOVE_START:
					editor.setSelection(0);
					break;
				case KEY_MOVE_END:
					editor.setSelection(editor.length());
					break;
				case KEY_MOVE_LEFT: {
					if (start > 0) {
						editor.setSelection(start - 1);
					}
					break;
				}
				case KEY_MOVE_RIGHT: {
					if (start < editor.length()) {
						editor.setSelection(start + 1);
					}
					break;
				}
				default: {
					if (!this.onKey(editor, primaryCode, keyCodes)) {
						if (editable != null) {
							editable.insert(editor.getSelectionStart(), Character.toString((char)primaryCode));
						}
					}
					// TODO consider the following:
					//long eventTime = SystemClock.uptimeMillis();
					//KeyEvent event = new KeyEvent(eventTime, eventTime, KeyEvent.ACTION_DOWN, primaryCode, 0, 0, 0, 0, KeyEvent.FLAG_SOFT_KEYBOARD | KeyEvent.FLAG_KEEP_TOUCH_MODE);
					//mTargetActivity.dispatchKeyEvent(event);
					break;
				}
			}
		}

		protected abstract boolean onKey(EditText editor, int primaryCode, int[] keyCodes);

		@Override
		public void onPress(int primaryCode) {
			if (haptic && primaryCode != 0) {
				keyboardView.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY);
			}
			//findEdit().onKeyDown(primaryCode, new KeyEvent(KeyEvent.ACTION_DOWN, primaryCode));
		}

		@Override
		public void onRelease(int primaryCode) {
			//findEdit().onKeyUp(primaryCode, new KeyEvent(KeyEvent.ACTION_UP, primaryCode));
		}

		@Override
		public void onText(CharSequence text) {
			EditText editor = findEdit();
			if (editor == null) {
				//? hideCustomKeyboard();
				return;
			}
			int start = editor.getSelectionStart();
			int end = editor.getSelectionEnd();
			editor.setText(text);
			editor.setSelection(start, end);
		}

		@Override
		public void swipeDown() {
		}

		@Override
		public void swipeLeft() {
		}

		@Override
		public void swipeRight() {
		}

		@Override
		public void swipeUp() {
		}
	}
}

File: feature/keyboard/src/main/kotlin/net/twisterrob/colorfilters/android/keyboard/BaseKeyboardHandler.kt
package net.twisterrob.colorfilters.android.keyboard

import android.annotation.SuppressLint
import android.content.Context
import android.text.Editable
import android.text.InputType
import android.util.Log
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import net.twisterrob.android.core.requireSystemService
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_BACKSPACE
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_CLEAR
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_DONE
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_MOVE_END
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_MOVE_LEFT
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_MOVE_RIGHT
import net.twisterrob.colorfilters.android.keyboard.KeyCodes.KEY_MOVE_START
import net.twisterrob.colorfilters.android.keyboard.KeyboardHandler.CustomKeyboardListener

// http://www.fampennings.nl/maarten/android/09keyboard/index.htm
// http://forum.xda-developers.com/showthread.php?t=2497237
abstract class BaseKeyboardHandler(
	protected val window: Window,
	@Suppress("DEPRECATION")
	protected val keyboardView: android.inputmethodservice.KeyboardView,
) : KeyboardHandler {

	private val context: Context = keyboardView.context.applicationContext

	override var hapticFeedback: Boolean = false
		set(value) {
			keyboardView.isHapticFeedbackEnabled = value
			field = value
		}

	override var customKeyboardListener: CustomKeyboardListener? = null

	init {
		hideCustomKeyboard()
		window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
		@Suppress("DEPRECATION")
		keyboardView.isPreviewEnabled = false
		hapticFeedback = true
	}

	override fun hideCustomKeyboard() {
		keyboardView.visibility = View.GONE
		keyboardView.isEnabled = false
		clearFocus()
		customKeyboardListener?.customKeyboardHidden()
	}

	private fun clearFocus() {
		window.currentFocus?.clearFocus()
	}

	override fun showCustomKeyboard(v: View) {
		val imm: InputMethodManager = context.requireSystemService() 
		imm.hideSoftInputFromWindow(v.windowToken, 0)
		keyboardView.isEnabled = true
		keyboardView.visibility = View.VISIBLE
		customKeyboardListener?.customKeyboardShown()
	}

	override fun registerEditText(editText: EditText) {
		// alternative to this is editText.setOnTouchListener(new OnTouchWrapper());
		try {
			val setShowSoftInputOnFocus =
				editText::class.java.getMethod("setShowSoftInputOnFocus", Boolean::class.java)
			setShowSoftInputOnFocus(editText, false)
		} catch (@Suppress("TooGenericExceptionCaught") ex: Exception) {
			// It's a hack, anything can go wrong.
			Log.w("HACK", "Could not turn off input focus for EditText: $editText", ex)
		}
		editText.isFocusable = true
		editText.isFocusableInTouchMode = true
		editText.onFocusChangeListener = MyOnFocusChangeListener()
		editText.setOnClickListener(MyOnClickListener())
	}

	override fun unregisterEditText(editText: EditText) {
		editText.onFocusChangeListener = null
		editText.setOnClickListener(null)
		editText.clearFocus()
	}

	override fun handleBack(): Boolean {
		if (@Suppress("DEPRECATION") keyboardView.handleBack()) {
			return true
		}
		if (keyboardView.visibility == View.VISIBLE) {
			hideCustomKeyboard()
			return true
		}
		return false
	}

	@Suppress("unused")
	@SuppressLint("ClickableViewAccessibility")
	private class OnTouchWrapper : View.OnTouchListener {

		override fun onTouch(v: View, event: MotionEvent): Boolean {
			val editText = v as EditText
			val backup = editText.inputType
			editText.inputType = InputType.TYPE_NULL // Disable standard keyboard
			try {
				return editText.onTouchEvent(event) // Call native handler
			} finally {
				editText.inputType = backup
				editText.isCursorVisible = true
			}
		}
	}

	private inner class MyOnClickListener : View.OnClickListener {

		override fun onClick(v: View) {
			showCustomKeyboard(v)
		}
	}

	private inner class MyOnFocusChangeListener : View.OnFocusChangeListener {

		override fun onFocusChange(v: View, hasFocus: Boolean) {
			if (hasFocus) {
				showCustomKeyboard(v)
			} else {
				hideCustomKeyboard()
			}
		}
	}

	@Suppress("OVERRIDE_DEPRECATION") // KeyboardView is deprecated, but the listener methods need to be used.
	protected abstract inner class BaseOnKeyboardActionListener 
		: @Suppress("DEPRECATION") android.inputmethodservice.KeyboardView.OnKeyboardActionListener {

		private fun findView(): View? {
			return window.currentFocus
		}

		private fun findEdit(): EditText? {
			return findView() as? EditText
		}

		@Suppress("ComplexMethod") // Curious if this can be improved, but not now.
		override fun onKey(primaryCode: Int, keyCodes: IntArray) {
			val editor = findEdit()
			if (editor == null) {
				hideCustomKeyboard()
				return
			}

			val editable: Editable? = editor.text

			// if there's a selection, clear it to make any edits replace actions
			val start = editor.selectionStart
			val end = editor.selectionEnd
			if (start < end) {
				requireNotNull(editable) { "Selection can't exist without text" }
					.delete(start, end)
			}

			when (primaryCode) {
				KEY_CLEAR -> editable?.clear()
				KEY_BACKSPACE ->
					if (editable != null && 0 < start) {
						editable.delete(start - 1, start)
					}
				KEY_DONE -> hideCustomKeyboard()
				KEY_MOVE_START -> editor.setSelection(0)
				KEY_MOVE_END -> editor.setSelection(editor.length())
				KEY_MOVE_LEFT ->
					if (start > 0) {
						editor.setSelection(start - 1)
					}
				KEY_MOVE_RIGHT ->
					if (start < editor.length()) {
						editor.setSelection(start + 1)
					}

				else -> {
					if (!this.onKey(editor, primaryCode, keyCodes)) {
						editable?.insert(editor.selectionStart, primaryCode.toChar().toString())
					}
					// TODO consider the following:
					//val eventTime = SystemClock.uptimeMillis()
					//val event = KeyEvent(eventTime, eventTime, KeyEvent.ACTION_DOWN, primaryCode, 0, 0, 0, 0, KeyEvent.FLAG_SOFT_KEYBOARD or KeyEvent.FLAG_KEEP_TOUCH_MODE)
					//mTargetActivity.dispatchKeyEvent(event)
				}
			}
		}

		protected abstract fun onKey(editor: EditText, primaryCode: Int, keyCodes: IntArray): Boolean

		override fun onPress(primaryCode: Int) {
			if (hapticFeedback && primaryCode != 0) {
				keyboardView.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
			}
			//findEdit().onKeyDown(primaryCode, KeyEvent(KeyEvent.ACTION_DOWN, primaryCode))
		}

		override fun onRelease(primaryCode: Int) {
			//findEdit().onKeyUp(primaryCode, KeyEvent(KeyEvent.ACTION_UP, primaryCode))
		}

		override fun onText(text: CharSequence?) {
			val editor = findEdit()
			@Suppress("FoldInitializerAndIfToElvis")
			if (editor == null) {
				//? hideCustomKeyboard()
				return
			}
			val start = editor.selectionStart
			val end = editor.selectionEnd
			editor.setText(text)
			editor.setSelection(start, end)
		}

		override fun swipeDown() {
			// Swiping gestures are not handled on these keyboards.
		}

		override fun swipeLeft() {
			// Swiping gestures are not handled on these keyboards.
		}

		override fun swipeRight() {
			// Swiping gestures are not handled on these keyboards.
		}

		override fun swipeUp() {
			// Swiping gestures are not handled on these keyboards.
		}
	}
}
